wasmcloud-chat/actors/messages/src/lib.rs

```diff
@@ -1,9 +1,172 @@
+extern crate serde;
+
 extern crate wapc_guest as guest;
-use wasmcloud_actor_core as actor;
+extern crate wasmcloud_actor_core as actor;
+extern crate wasmcloud_actor_eventstreams as streams;
+extern crate wasmcloud_actor_messaging as broker;
+extern crate wasmcloudchat_messages_interface as messages;
 
+use chrono::prelude::*;
 use guest::prelude::*;
+use messages::{ChannelMessage, ProcessAck};
+use std::collections::HashMap;
+use url::Url;
+
+mod cloudevent;
+
+use cloudevent::CloudEvent;
+
+const WCCHAT_SCHEME: &str = "wcchat";
+const EVENT_TYPE_MESSAGE_PUBLISHED: &str = "com.wasmcloud.chat.events.messagepublished";
+const EVENT_SOURCE: &str = "/actors/messaging";
+
+const HMK_ORIGIN_CHANNEL: &str = "origin_channel";
+const HMK_ORIGIN_USER: &str = "origin_user";
+const HMK_TIMESTAMP: &str = "timestamp";
+const HMK_MESSAGE_TEXT: &str = "messagetext";
+const HMK_ID: &str = "message_id";
+
+const MSGTYPE_MESSAGE: &str = "MSG";
+const VALID_MESSAGE_TYPES: [&str; 1] = [MSGTYPE_MESSAGE];
 
 #[actor::init]
 fn init() {
-    // Register your message handlers here
-}
\ No newline at end of file
+    messages::Handlers::register_process_message(process_message);
+}
+
+// Processing a message:
+// 1 - validate the message
+// 2 - publish it to the appropriate message broker subject
+// 3 - emit event to event stream
+// 4 - ACK
+fn process_message(message: ChannelMessage) -> HandlerResult<ProcessAck> {
+    let target = MessageTarget::from(&message);
+    if let MessageTarget::Unknown(message) = target {
+        return Err(format!("Unable to select target for message: {}", message).into());
+    }
+
+    Ok(validate_message(&message)
+        .and_then(|_| publish_broker_message(&target, &message))
+        .and_then(|_| emit_stream_event(&target, &message))
+        .map_or_else(
+            |_| ProcessAck::success(&message.message_id),
+            |_| {
+                ProcessAck::fail(
+                    &message.message_id,
+                    "Did not publish and emit to event stream",
+                )
+            },
+        ))
+}
+
+fn validate_message(message: &messages::ChannelMessage) -> HandlerResult<()> {
+    if message.message_id.is_empty() {
+        return Err("No message ID".into());
+    }
+    if message.target_url == format!("wcchat://users/{}", message.origin_user_id) {
+        return Err("Cannot send messages to yourself".into());
+    }
+    if message.message_text.is_empty() {
+        return Err("Cannot send empty messages".into());
+    }
+    if !VALID_MESSAGE_TYPES
+        .to_vec()
+        .contains(&message.message_type.as_str())
+    {
+        return Err("Invalid/unrecognized message type".into());
+    }
+    Ok(())
```

Comment:
Is there a reason that this isn't a single if/elseif/else statement? Since all of these conditions are tested individually and sequentially, the `return` statements could be omitted

Reply:
This is the "go" style of handling function-aborting conditionals. It's really a matter of preference, but I prefer to see a list of "if this then bail" statements than the slightly harder to decipher if/else if/else where the "happy path" is actually inside an else.

Reply:
Looks very similar to the way that `Go` handles things, works for me 👍🏻